clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1666,13 +1668,28 @@ private boolean initWithCommittedOffsetsIfNeeded(Timer timer) {
             return true;
 
         log.debug("Refreshing committed offsets for partitions {}", initializingPartitions);
+
+        // The shorter the timeout provided to poll(), the more likely the offsets fetch will time out. To handle
+        // this case, on the first attempt to fetch the committed offsets, a FetchCommittedOffsetsEvent is created
+        // (with potentially a longer timeout) and stored. The event is used for the first attempt, but in the
+        // case it times out, subsequent attempts will also use the event in order to wait for the results.
+        if (!canReusePendingOffsetFetchEvent(initializingPartitions)) {
+            // Give the event a reasonable amount of time to complete.
+            long timeoutMs = Math.max(defaultApiTimeoutMs, timer.remainingMs());
+            long deadlineMs = time.milliseconds() + timeoutMs;
+            pendingOffsetFetchEvent = new FetchCommittedOffsetsEvent(initializingPartitions, deadlineMs);
+            applicationEventHandler.add(pendingOffsetFetchEvent);
+        }
+
+        final CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> future = pendingOffsetFetchEvent.future();
+
         try {
-            final FetchCommittedOffsetsEvent event =
-                new FetchCommittedOffsetsEvent(
-                    initializingPartitions,
-                    calculateDeadlineMs(timer));
-            wakeupTrigger.setActiveTask(event.future());
-            final Map<TopicPartition, OffsetAndMetadata> offsets = applicationEventHandler.addAndGet(event);
+            wakeupTrigger.setActiveTask(future);
+            final Map<TopicPartition, OffsetAndMetadata> offsets = ConsumerUtils.getResult(future, timer);
```

Comment:
The approach looks good but `getResult` can throw exceptions. There are three cases to consider I think:

* TimeoutException - This is the case this PR is aiming to fix, so I think this is fine.
* InterruptedException - I think you can also leave the pending event in place.
* Everything else - Probably ought to clear the pendingOffsetFetchEvent because the event is doomed.

Reply:
I want to clarify the desired error handling logic:

- `TimeoutException`: log message, keep pending event, return `false`
- `InterruptException`: no log, keep pending event, throws error
- Everything else - no log, clears pending event, throws `KafkaException`-wrapped error

Reply:
Is there any _harm_ in having `InterruptException` clear out the pending event? Why else would an interrupt exception be thrown in normal operations?

Reply:
There's no _harm_ in having `InterruptException` clear out the pending event.

Reply:
_HARM_